### DISCUSSION TITLE
Guide: How to build MSIX for Microsoft Store (WASDK 1.3)

### DISCUSSION THREAD

**Discussion by omikhailov** [microsoft/microsoft-ui-xaml#8472] [Show and tell]
Congratulate me, I just submitted my first WinUI 3 app to the store. So far, the technology makes a contradictory impression, and the most negative contribution to this assessment is made by the MSIX package assembling stage.

At the moment, Visual Studio does not make valid MSIX packages. Well, you can click Package and Publish -> Create App Packages just like you used to do before for UWP apps, but if you submit generated packages, they will fail certification in the Microsoft Store.

Problem is that WinUI 3 app depends on the Windows App SDK Runtime and .Net 6 Runtime, but Visual Studio does not handle these dependencies correctly.

The first one is easy to fix, because Windows App SDK is available in the Microsoft Store as framework package and all you need is just to add a reference to that package in your **Package.appxmanifest**:

```xml
  <Dependencies>
    <TargetDeviceFamily Name="Windows.Universal" MinVersion="10.0.17763.0" MaxVersionTested="10.0.19041.0" />
    <TargetDeviceFamily Name="Windows.Desktop" MinVersion="10.0.17763.0" MaxVersionTested="10.0.19041.0" /> 

    <PackageDependency Name="Microsoft.WindowsAppRuntime.1.3" Publisher="CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US" MinVersion="3000.820.152.0" />
  </Dependencies>
```

Looks simply, but I spent hours looking up the name, publisher and version for the package - it is just not documented. If in the future you need to do something similar, open a PowerShell and use following command to search for framework packages installed on your machine:

`Get-AppxPackage | Where-Object IsFramework`

Unlike the App SDK, .Net 6 Runtime is not published in the Store and Microsoft does not preinstall it with Windows, So, if you will publish your app now, it will show an ugly dialog box asking users to install .Net on the first launch. And that's not all, Microsoft Store team will reject your submission because they have a rule claiming that if your app behaves this way, you must add a notice about such dependencies in the first two lines of the app description. Like, "Hey, this is MyCoolAppName that requires .Net SDK version 6.666 dot zero five six seven... Please don't be afraid to download it".

To do not spoil the user experience, you have to include whole .Net 6 runtime into your package. It's size will increase from 9 megabytes to 40, and Microsoft Store will report that your app requires almost 100 megabytes on the disk, but there is no other way, unfortunatelly.

So, open **.csproj** file of your app and add options for self-contained app:

```xml
<?xml version="1.0" encoding="utf-8"?>
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>net6.0-windows10.0.19041.0</TargetFramework>
    <TargetPlatformMinVersion>10.0.17763.0</TargetPlatformMinVersion>
    <RootNamespace>ThreadMonitor</RootNamespace>
    <ApplicationManifest>app.manifest</ApplicationManifest>
    <Platforms>x86;x64;ARM64</Platforms>
    <RuntimeIdentifiers>win10-x86;win10-x64;win10-arm64</RuntimeIdentifiers>
    <PublishProfile>win10-$(Platform).pubxml</PublishProfile>
    <UseWinUI>true</UseWinUI>
    <EnableMsixTooling>true</EnableMsixTooling>
    <GenerateTemporaryStoreCertificate>True</GenerateTemporaryStoreCertificate>

    <PublishSingleFile>false</PublishSingleFile>
    <SelfContained>true</SelfContained>
  </PropertyGroup>
```
Now, if you will try to package the app, Visual Studio may fail with error related to platforms mismatch or something like that. I think that in my case it happened because I had more than one project in a solution, but it could be also an issue in the "Blank App, Packaged (WinUI 3 Desktop)" template itself. Anyway, if it will happen to you, add also following sections into your .csproj file:

```xml
<?xml version="1.0" encoding="utf-8"?>
<Project Sdk="Microsoft.NET.Sdk">
...
    <PropertyGroup Condition="'$(Configuration)' == 'Release' and $(PlatformTarget)=='x64'">		
	<RuntimeIdentifier>win10-x64</RuntimeIdentifier>
    </PropertyGroup>
    <PropertyGroup Condition="'$(Configuration)' == 'Release' and $(PlatformTarget)=='x86'">
	<RuntimeIdentifier>win10-x86</RuntimeIdentifier>
    </PropertyGroup>
    <PropertyGroup Condition="'$(Configuration)' == 'Release' and $(PlatformTarget)=='ARM64'">
	<RuntimeIdentifier>win10-arm64</RuntimeIdentifier>
    </PropertyGroup>
```

Perhaps, you understood an idea - since Visual Studio cannot match platforms and runtime identifiers, you have to map them manually for all platforms you are going to target.

In addition, open Configuration Manager and check that all your projects have the same targets. In my case, a library project had only "Any CPU" configuration available, so I manually edited .sln file to add them:

```
Microsoft Visual Studio Solution File, Format Version 12.00
# Visual Studio Version 17
VisualStudioVersion = 17.4.33213.308
MinimumVisualStudioVersion = 10.0.40219.1
Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Thread Monitor", "Thread Monitor\Thread Monitor.csproj", "{9E4F7E28-A7E1-4966-8938-CD1937E25D9E}"
EndProject
Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "ThreadMonitor.BL", "ThreadMonitor.BL\ThreadMonitor.BL.csproj", "{EAF2CA36-A554-47D9-A7E0-0284174E64E4}"
EndProject
Global
	GlobalSection(SolutionConfigurationPlatforms) = preSolution
		Debug|ARM64 = Debug|ARM64
		Debug|x64 = Debug|x64
		Debug|x86 = Debug|x86
		Release|ARM64 = Release|ARM64
		Release|x64 = Release|x64
		Release|x86 = Release|x86
	EndGlobalSection
	GlobalSection(ProjectConfigurationPlatforms) = postSolution
		{9E4F7E28-A7E1-4966-8938-CD1937E25D9E}.Debug|ARM64.ActiveCfg = Debug|ARM64
		{9E4F7E28-A7E1-4966-8938-CD1937E25D9E}.Debug|ARM64.Build.0 = Debug|ARM64
		{9E4F7E28-A7E1-4966-8938-CD1937E25D9E}.Debug|ARM64.Deploy.0 = Debug|ARM64
		{9E4F7E28-A7E1-4966-8938-CD1937E25D9E}.Debug|x64.ActiveCfg = Debug|x64
		{9E4F7E28-A7E1-4966-8938-CD1937E25D9E}.Debug|x64.Build.0 = Debug|x64
		{9E4F7E28-A7E1-4966-8938-CD1937E25D9E}.Debug|x64.Deploy.0 = Debug|x64
		{9E4F7E28-A7E1-4966-8938-CD1937E25D9E}.Debug|x86.ActiveCfg = Debug|x86
		{9E4F7E28-A7E1-4966-8938-CD1937E25D9E}.Debug|x86.Build.0 = Debug|x86
		{9E4F7E28-A7E1-4966-8938-CD1937E25D9E}.Debug|x86.Deploy.0 = Debug|x86
		{9E4F7E28-A7E1-4966-8938-CD1937E25D9E}.Release|ARM64.ActiveCfg = Release|ARM64
		{9E4F7E28-A7E1-4966-8938-CD1937E25D9E}.Release|ARM64.Build.0 = Release|ARM64
		{9E4F7E28-A7E1-4966-8938-CD1937E25D9E}.Release|ARM64.Deploy.0 = Release|ARM64
		{9E4F7E28-A7E1-4966-8938-CD1937E25D9E}.Release|x64.ActiveCfg = Release|x64
		{9E4F7E28-A7E1-4966-8938-CD1937E25D9E}.Release|x64.Build.0 = Release|x64
		{9E4F7E28-A7E1-4966-8938-CD1937E25D9E}.Release|x64.Deploy.0 = Release|x64
		{9E4F7E28-A7E1-4966-8938-CD1937E25D9E}.Release|x86.ActiveCfg = Release|x86
		{9E4F7E28-A7E1-4966-8938-CD1937E25D9E}.Release|x86.Build.0 = Release|x86
		{9E4F7E28-A7E1-4966-8938-CD1937E25D9E}.Release|x86.Deploy.0 = Release|x86

#Library project configuration was incorrect, so I copied configurations from {9E4F7E28-A7E1-4966-8938-CD1937E25D9E}

		{EAF2CA36-A554-47D9-A7E0-0284174E64E4}.Debug|ARM64.ActiveCfg = Debug|ARM64
		{EAF2CA36-A554-47D9-A7E0-0284174E64E4}.Debug|ARM64.Build.0 = Debug|ARM64
		{EAF2CA36-A554-47D9-A7E0-0284174E64E4}.Debug|ARM64.Deploy.0 = Debug|ARM64
		{EAF2CA36-A554-47D9-A7E0-0284174E64E4}.Debug|x64.ActiveCfg = Debug|x64
		{EAF2CA36-A554-47D9-A7E0-0284174E64E4}.Debug|x64.Build.0 = Debug|x64
		{EAF2CA36-A554-47D9-A7E0-0284174E64E4}.Debug|x64.Deploy.0 = Debug|x64
		{EAF2CA36-A554-47D9-A7E0-0284174E64E4}.Debug|x86.ActiveCfg = Debug|x86
		{EAF2CA36-A554-47D9-A7E0-0284174E64E4}.Debug|x86.Build.0 = Debug|x86
		{EAF2CA36-A554-47D9-A7E0-0284174E64E4}.Debug|x86.Deploy.0 = Debug|x86
		{EAF2CA36-A554-47D9-A7E0-0284174E64E4}.Release|ARM64.ActiveCfg = Release|ARM64
		{EAF2CA36-A554-47D9-A7E0-0284174E64E4}.Release|ARM64.Build.0 = Release|ARM64
		{EAF2CA36-A554-47D9-A7E0-0284174E64E4}.Release|ARM64.Deploy.0 = Release|ARM64
		{EAF2CA36-A554-47D9-A7E0-0284174E64E4}.Release|x64.ActiveCfg = Release|x64
		{EAF2CA36-A554-47D9-A7E0-0284174E64E4}.Release|x64.Build.0 = Release|x64
		{EAF2CA36-A554-47D9-A7E0-0284174E64E4}.Release|x64.Deploy.0 = Release|x64
		{EAF2CA36-A554-47D9-A7E0-0284174E64E4}.Release|x86.ActiveCfg = Release|x86
		{EAF2CA36-A554-47D9-A7E0-0284174E64E4}.Release|x86.Build.0 = Release|x86
		{EAF2CA36-A554-47D9-A7E0-0284174E64E4}.Release|x86.Deploy.0 = Release|x86
	EndGlobalSection
	GlobalSection(SolutionProperties) = preSolution
		HideSolutionNode = FALSE
	EndGlobalSection
	GlobalSection(ExtensibilityGlobals) = postSolution
		SolutionGuid = {D3297F57-B284-4EFC-9DF2-C0C9F69AAA80}
	EndGlobalSection
EndGlobal
```

Finally! Valid MSIX package and the cherished letter from Microsoft

![image](https://github.com/microsoft/microsoft-ui-xaml/assets/75426711/d6b66a02-3532-415b-815c-77b0b0c0a311)

**Comment by lhak** [comment]
Have you checked that the publish profiles are present in the Properties\PublishProfiles folder? They are included in the .csproj file and should set properties like SelfContained and RuntimeIdentifier to the correct settings.

**Reply by omikhailov** [reply]
> Have you checked that the publish profiles are present in the Properties\PublishProfiles folder? They are included in the .csproj file and should set properties like SelfContained and RuntimeIdentifier to the correct settings.

They are needed for publishing unpackaged apps, I'm talking here only about packaged.

**Reply by lhak** [reply]
I just tested it and the publish profiles are always evaluated when you run an application from Visual Studio or publish it, both in unpackaged and packaged configuration.
